src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs

```diff
@@ -710,6 +708,50 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
             }
 
             return UpdateExpression(builder, node.Update(left, right, node.IsRef, node.Type));
+
+            BoundExpression spillReceiver(BoundFieldAccess field, ref BoundSpillSequenceBuilder leftBuilder, bool isFieldLeft)
+            {
+                BoundExpression receiver;
+                if (field.FieldSymbol.ContainingType.IsReferenceType)
+                {
+                    // a reference type can always live across await so Spill using leftBuilder
+                    receiver = Spill(leftBuilder, VisitExpression(ref leftBuilder, field.ReceiverOpt));
+                }
+                else if (field.ReceiverOpt is BoundArrayAccess arrayAccess)
+                {
+                    // an arrayAccess returns a ref so can only be called after the await, but spill expression and indices
+                    var expression = VisitExpression(ref leftBuilder, arrayAccess.Expression);
+                    expression = Spill(leftBuilder, expression, RefKind.None);
+                    var indices = this.VisitExpressionList(ref leftBuilder, arrayAccess.Indices, forceSpill: true);
+                    receiver = arrayAccess.Update(expression, indices, arrayAccess.Type);
+                    // dummy array access to trigger IndexOutRangeException or NRE
+                    // we only need this if the array access is a receiver since
+                    // a[0] = b triggers a NRE/IORE on assignment
+                    // but a[0].b = c triggers an NRE/IORE before evaluating c
+                    Spill(leftBuilder, receiver, sideEffectsOnly: true);
+                }
+                else if (field.ReceiverOpt is BoundFieldAccess receiverField)
+                {
+                    receiver = spillReceiver(receiverField, ref leftBuilder, isFieldLeft: false);
+                }
+                else
+                {
+                    receiver = Spill(leftBuilder, VisitExpression(ref leftBuilder, field.ReceiverOpt), RefKind.Ref);
+                }
+
+                field = field.Update(receiver, field.FieldSymbol, field.ConstantValueOpt, field.ResultKind, field.Type);
+
+                // a.b = c will trigger a NRE if a is null on assignment,
+                // but a.b.c = d will trigger a NRE if a is null before evaluating d
+                // so check whether we assign to the field directly
+                if (!isFieldLeft && field.FieldSymbol.ContainingType.IsReferenceType)
```

Comment:
>if (!isFieldLeft && field.FieldSymbol.ContainingType.IsReferenceType) [](start = 16, length = 69)

It feels like the logic around evaluating exception side-effects for the receiver is somewhat confusing. For example, it looks like the only way we get inside this ```if``` is when we get into the ```if (field.FieldSymbol.ContainingType.IsReferenceType)``` at the top of this local function. Why don't we move this code there? I think that would make it pretty clear that exception side-effects are evaluated only once.